### PR TITLE
fix: isolate conditional-logit RNG from global NumPy state (#193)

### LINKED ISSUE #193: Replace global `np.random.seed` in choice models with a local Generator
## Summary

Replace the two process-global `np.random.seed(random_state)` calls in the
conditional-logit choice model with a local `numpy.random.Generator`, matching
the `default_rng(...)` pattern already used elsewhere in the codebase.

## Why this matters

Calling `np.random.seed` mutates the global NumPy RNG for the whole process. A
user who runs an evaluation and then does their own sampling gets silently
reseeded state; concurrent evaluations interfere; and the behaviour is
inconsistent with `_simulation.py` and `scenarios.py`, which already use
`np.random.default_rng(...)`. Removing global side effects strengthens
reproducibility — a first-order concern for a statistical-evaluation library.

## Current evidence

- `src/skdr_eval/choice.py` line 60 and line 263: `np.random.seed(random_state)`
  followed by `np.random.normal(...)`.
- Contrast: `src/skdr_eval/_simulation.py:188` and `src/skdr_eval/scenarios.py:63`
  use `np.random.default_rng(...)` (the modern, side-effect-free API).

## External context

NumPy's own guidance recommends `default_rng()` Generators over the legacy
global `np.random.seed`/`np.random.*` functions for reproducible, isolated
randomness. Not strictly required to cite in the issue body.

## Proposed implementation

1. In `choice.py`, construct `rng = np.random.default_rng(random_state)` at the
   top of each affected function.
2. Replace `np.random.normal(...)` with `rng.normal(...)`.
3. Thread the generator through any helper that currently relies on global state.
4. Confirm the conditional-logit fit is deterministic given a fixed
   `random_state` and unaffected by external global reseeds.

## AI-agent execution notes

- Inspect first: `src/skdr_eval/choice.py`, `tests/test_propensity_choice.py`.
- This touches statistical logic indirectly (initialization of the optimizer):
  per `docs/agent-context/review-checklist.md`, include a simulation/test showing
  the conditional-logit coefficients still recover a known ground-truth choice
  probability under the new Generator, and that results are stable.
- Edge cases: `random_state=None`, integer, and `np.random.Generator` inputs.
- Do not change the optimizer, the objective, or default values.

## Acceptance criteria

- No `np.random.seed(...)` call remains in `src/skdr_eval/choice.py`.
- A test asserts that an unrelated `np.random.seed(...)` before/after a fit does
  not change the fitted coefficients.
- The existing recovery/equivalence tests for the choice model still pass.

## Test plan

- Unit: determinism under fixed `random_state`; isolation from global reseed.
- Simulation: recover a known conditional-logit ground-truth coefficient.
- CI: existing `tests/test_propensity_choice.py` + `make validate`.

## Documentation plan

- CHANGELOG `### Changed` (or `### Fixed`) entry noting the RNG isolation; clarify
  it does not change default behaviour given the same `random_state`.

## Migration and compatibility notes

Numerically the fit may differ at the last ULPs because the draw sequence from a
Generator differs from the legacy global stream. Treat as a numerical-behaviour
note in CHANGELOG; provide the one-line statement that results remain
deterministic per `random_state`.

## Risks and tradeoffs

Minor: the exact random initialization sequence changes, so a pinned numeric
snapshot (if any) shifts. Alternative: keep global seed but save/restore state —
rejected as more fragile than a local Generator.

## Suggested labels

reliability, refactor, testing, good-first-ai-issue

### PR DESCRIPTION
**Pull request by dgenio**
# Pull Request

## 📋 Description

Replaces the two process-global `np.random.seed(random_state)` calls in `src/skdr_eval/choice.py` with a local `np.random.default_rng(...)` Generator, matching the `default_rng` pattern already used in `_simulation.py` and `scenarios.py`. Fitting the conditional-logit model no longer silently reseeds the caller's global NumPy RNG or interferes with concurrent evaluations — a first-order reproducibility concern for a statistical-evaluation library.

`random_state` now also accepts an `np.random.Generator` or `None` in addition to an integer seed.

### Type of Change
- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [x] 🧹 Code refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [x] 🧪 Test improvements
- [ ] 🔧 Build/CI improvements

## 🔗 Related Issues
- Fixes #193

### Test Coverage
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have added integration tests if applicable — N/A (unit-level RNG isolation)

### Manual Testing
- [x] Tested locally (`python -m pytest`, `ruff`, `mypy`)
- [x] Tested with example scripts (preflight + quickstart smoke)
- [x] Tested edge cases (`random_state` as int, `Generator`, and `None`)

**Test commands run:**
```bash
ruff check src/ tests/ examples/            # All checks passed
ruff format --check src/ tests/ examples/   # All formatting OK
mypy src/skdr_eval/                          # Success: no issues in 44 files
python -m pytest -q --cov=skdr_eval          # 948 passed, 9 skipped — 93.27% coverage
```

> Note: `make validate` runs the bare `pytest` console script, which in this sandbox resolves to a separate uv-tool install lacking the project's deps (`matplotlib`); the faithful run is `python -m pytest`, which is green. This is a pre-existing environment quirk, not introduced by this PR.

### Code Changes
- `src/skdr_eval/choice.py`: `fit_conditional_logit`, `sample_negative_pairs`, and `fit_conditional_logit_with_sampling` now build a local `np.random.default_rng(random_state)` and draw via `rng.normal` / `rng.choice` instead of `np.random.seed(...)` + `np.random.*`. Added a `RandomStateLike` type alias (`int | np.random.Generator | None`) and updated docstrings.
- `tests/test_propensity_choice.py`: added determinism, global-reseed isolation, no-global-mutation (the regression target), edge-case seed types, and a ground-truth recovery simulation; plus a `_make_condlogit_data` helper.
- `CHANGELOG.md`: `### Changed` entry under Unreleased ([#193]).

### API Changes
- [ ] No API changes
- [x] Backward compatible API additions — `random_state` widened from `int` to `int | np.random.Generator | None`; existing integer callers are unaffected.
- [ ] Breaking API changes (requires major version bump)

## 📚 Documentation
- [x] I have updated the documentation accordingly — docstrings updated; no public quickstart/example references these internal functions, so no `examples/`/README change needed.
- [x] I have updated docstrings for new/modified functions
- [ ] I have added examples if applicable — N/A
- [x] I have updated the CHANGELOG.md

### Code Quality
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have run `ruff check` and `ruff format`
- [x] I have run `mypy` type checking

### Testing & CI
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes
- [ ] All CI checks pass — pending CI
- [x] Code coverage is maintained or improved (93.27%)

### Documentation & Communication
- [x] I have made corresponding changes to the documentation
- [x] My commit messages follow the conventional commit format
- [x] I have updated the CHANGELOG.md

### Focus Areas
- The key regression test is `test_fit_conditional_logit_does_not_mutate_global_rng` — it fails against the old `np.random.seed` behaviour and passes now (verified).
- Simulation proof (`test_fit_conditional_logit_recovers_ground_truth`) recovers `beta=[1.5, -1.0]` to within max abs error ≈0.044 (`atol=0.2`), satisfying the repo's Definition of Done for statistical-logic changes.

### Note on threading
In `fit_conditional_logit_with_sampling`, passing an integer seeds each step reproducibly; passing a `Generator` threads a single stream through both negative sampling and optimizer init (no draw reuse).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01AEGSrSJ3abcoGCnmyQNva7)_

### CODE REVIEW

File: src/skdr_eval/choice.py [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
 import logging
+from typing import TypeAlias
 
 import numpy as np
 
 logger = logging.getLogger("skdr_eval")
 
+# Accepted seed types for the local ``numpy`` Generator. Mirrors what
+# ``np.random.default_rng`` accepts: an integer seed, an existing Generator
+# (passed through unchanged), or ``None`` for non-deterministic seeding.
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/choice.py]: Good catch — reworded in 8b38c45. The comment now states this annotation covers the subset of seed types the API documents/tests (int, Generator, None), and notes that `default_rng` also accepts other forms (SeedSequence, BitGenerator, array-like) which still work at runtime.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_propensity_choice.py [hunk 2]
```diff
@@ -284,5 +308,149 @@ def test_large_dataset_fallback():
...
+        pytest.skip("SciPy not installed in this environment")
+
+    x, choice_ids, y = _make_condlogit_data(
+        n_sets=400, k_alts=4, beta=[1.0, -0.5], rng=np.random.default_rng(0)
+    )
+    coef_a, int_a, _ = fit_conditional_logit(x, choice_ids, y, random_state=42)
+    coef_b, int_b, _ = fit_conditional_logit(x, choice_ids, y, random_state=42)
+
+    np.testing.assert_array_equal(coef_a, coef_b)
+    assert int_a == int_b
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_propensity_choice.py]: Switched to `assert_allclose(rtol=0, atol=1e-12)` in 8b38c45. Worth noting these two fits run the identical computation in the same process (same BLAS), so they're bitwise-identical rather than cross-platform — but the tight tolerance still catches any RNG/determinism regression (which moves the optimizer trajectory by far more than 1e-12) and avoids float `==` as a code smell. Left the RNG-stream and integer-index comparisons as exact, since those are reproducible by definition.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_propensity_choice.py [hunk 3]
```diff
@@ -284,5 +308,149 @@ def test_large_dataset_fallback():
...
+        n_sets=400, k_alts=4, beta=[1.0, -0.5], rng=np.random.default_rng(1)
+    )
+
+    np.random.seed(0)
+    coef_a, int_a, _ = fit_conditional_logit(x, choice_ids, y, random_state=42)
+    np.random.seed(123456)  # would change a global-RNG initialization
+    coef_b, int_b, _ = fit_conditional_logit(x, choice_ids, y, random_state=42)
+
+    np.testing.assert_array_equal(coef_a, coef_b)
+    assert int_a == int_b
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_propensity_choice.py]: Done in 8b38c45 — same change here, `assert_allclose(rtol=0, atol=1e-12)`. Any leak of global state into the fit would shift the coefficients far more than the tolerance, so isolation is still verified.

---
_Generated by [Claude Code](https://claude.ai/code)_
